### PR TITLE
boards/arm: Adding Contextual Electronics Advanced BLE Cell board to Zephyr

### PR DESCRIPTION
**Pull request by bwasim**
This commit adds supports for the nRF52840 based BLE Cell board from Contextual
Electronics. This board contains support for BG95 Modem, BQ52895 charger, SD
card etc and can be used as a PI Hat.

In this commit, this board supports UART, I2C, SPI, Modem. Support for charger,
SD card and other things will be added later.

Signed-off-by: Bilal Wasim <bilalwasim676@gmail.com>

### CODE REVIEW

**Comment by bwasim** [comment]
cc: @carlescufi

File: boards/arm/contextualelectronics_abc/contextualelectronics_abc.dts
Comment (parthitce) [inline boards/arm/contextualelectronics_abc/contextualelectronics_abc.dts:91]: This page is removed. This comment can be removed.

Comment (bwasim) [reply boards/arm/contextualelectronics_abc/contextualelectronics_abc.dts:91]: Thanks for pointing out, Will remove.

**Comment by bwasim** [comment]
@carlescufi , This is a custom board based on nRF52840 SoC. Can I get a review from you and other Nordic maintainers !

File: boards/arm/contextualelectronics_abc/doc/index.rst
Comment (pabigot) [inline boards/arm/contextualelectronics_abc/doc/index.rst:76]: It's not obvious from  https://contextualelectronics.com/ where this board can be found; a link to https://contextualelectronics.com/courses/advanced-ble-cell/ might be more helpful.

Comment (bwasim) [reply boards/arm/contextualelectronics_abc/doc/index.rst:76]: Fixed !

File: boards/arm/contextualelectronics_abc/board.cmake
Comment (mbolivar-nordic) [inline boards/arm/contextualelectronics_abc/board.cmake:4]: This line is not needed anymore since 6628a16e4d7aa46e57fd2cc8796592983bcc6516; please remove.

I've posted a separate PR (https://github.com/zephyrproject-rtos/zephyr/pull/31121) eliminating it from elsewhere in the tree.

File: boards/arm/contextualelectronics_abc/contextualelectronics_abc.dts
Comment (mbolivar-nordic) [inline boards/arm/contextualelectronics_abc/contextualelectronics_abc.dts:75]: If that's the case, why isn't there a subnode for it here? Even if there's no driver, it seems convenient to have a binding and a node, to be able to get its address etc.

File: boards/arm/contextualelectronics_abc/doc/index.rst
Comment (mbolivar-nordic) [inline boards/arm/contextualelectronics_abc/doc/index.rst:125]: Looks like this is copy/pasted from elsewhere in the docs, but I don't think it should be here; it's just duplicating information in https://docs.zephyrproject.org/latest/guides/dts/howtos.html#set-devicetree-overlays (and slightly inaccurately, at that); please remove or replace with a link to that doc.

I've sent a separate PR  (https://github.com/zephyrproject-rtos/zephyr/pull/31123) fixing the others.

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
I gave this a quick once-over and it generally seems OK. A couple of docs requests and some minor questions. Interesting course idea, glad to see Zephyr there.

File: boards/arm/contextualelectronics_abc/contextualelectronics_abc.dts
Comment (bwasim) [reply boards/arm/contextualelectronics_abc/contextualelectronics_abc.dts:75]: I'm working on upstreaming the bq2589x driver for zephyr (https://github.com/bwasim/zephyr/tree/bq25895-zephyr), And the bindings / driver haven't been finalized yet. I will remove this comment for now and add the device tree node whenever I upstream the driver.

**Comment by bwasim** [comment]
Getting a strange buildkite failure which is unrelated to the pull request (https://buildkite.com/zephyr/zephyr/builds/17251#31c9b051-7f2e-4171-b824-0aed10314ab2). Seems that the server node is out of memory.

**Review by pabigot — APPROVED** [review]
Seems reasonable.

My only concern is that this board is being used in a class on designing a board, and the documentation states:

> This will go all the way through production, as this board will be used as lead generation for a consulting business and also possibly being put up for sale as a development board other people can use to design cellular and bluetooth devices.

How likely is it this board will be made available?

**Comment by bwasim** [comment]
> How likely is it this board will be made available?

This would be a question for @ChrisGammell

**Comment by ChrisGammell** [comment]
> > How likely is it this board will be made available?
> 
> This would be a question for @ChrisGammell

The initial plan for the board was the educational aspect, but I have since decided to put the boards up for sale for others. I'm still working through the details of that.

I will update the documentation to reflect future availability.

File: boards/arm/contextualelectronics_abc/doc/index.rst
Comment (mbolivar-nordic) [inline boards/arm/contextualelectronics_abc/doc/index.rst:86]: Thanks for the update.

Taking a second look at this, that's not really true. The board.cmake file has all the necessary boilerplate to set up debugging.

I looked at the block diagram and I think the issue is just that you need an external debug probe. Perhaps just say that directly instead?

Comment (bwasim) [reply boards/arm/contextualelectronics_abc/doc/index.rst]: Good catch ! This was a copy-paste from the nRF52840 DK doc. I've refactored it to look more similar to the adafruit_feather_nrf52840 readme.

Comment (mbolivar-nordic) [reply boards/arm/contextualelectronics_abc/doc/index.rst]: >  This was a copy-paste from the nRF52840 DK doc.

Ugh, which is also wrong ... :D. I'll fix that too, thanks for letting me know.

**Review by mbolivar-nordic — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**
